Python/chapter03/p04_queue_via_stacks/miguelHx.py

```diff
@@ -0,0 +1,290 @@
+            IndexError: [description]
+        Returns:
+            int: the value at the top of the stack
+        """
+        if self.top is None:
+            raise IndexError('Stack is Empty')
+        return self.top.data
+    
+    def empty(self) -> None:
+        while self._size > 0:
+            self.pop()
+
+    def __iter__(self) -> Iterator:
+        """
+        Builds a list of the current stack state.
+        For example, given the following stack:
+            3 -> 2 -> 1, where 3 is the top,
+        Expect:
+            [3, 2, 1]
+        Returns:
+            List[int]: list of integers
+        """
+        return self
+
+    def __next__(self) -> T:
+        self.index += 1
+        if self.index == self._size or self.current_node is None:
+            self.index = -1
+            self.current_node = self.top
+            raise StopIteration
+        n: T = self.current_node.data
+        self.current_node = self.current_node.next
+        return n
+
+    def __bool__(self) -> bool:
+        """
+        True is returned when the container is not empty.
+        From https://docs.python.org/3/reference/datamodel.html#object.__bool__ :
+        Called to implement truth value testing and the built-in operation bool();
+        should return False or True. When this method is not defined, len() is called,
+        if it is defined, and the object is considered true if its result is nonzero.
+        If a class defines neither len() nor bool(), all its instances are considered true.
+        Returns:
+            bool: False when empty, True otherwise
+        """
+        return self._size > 0
+    
+    def __len__(self) -> int:
+        return self._size
+
+    def __str__(self):
+        if self._size == 0:
+            return '<Empty>'
+        values = []
+        n = self.top
+        while n.next is not None:
+            values.append(str(n.data))
+            n = n.next
+        values.append(str(n.data))
+        return '->'.join(values)
+
+
+@dataclass
+class QueueNode(Generic[T]):
+    data: T
+    next: 'Optional[QueueNode[T]]'
+
+
+class MyQueue:
+    def __init__(self):
+        self._size = 0
+        self.stack_one = MyStack()  # will be used as main data container
+        self.stack_two = MyStack()  # will be used to extract the least recent insertion
+    
+    def add(self, item: T):
+        self.stack_one.push(item)
+        self._size += 1
+    
+    def remove(self) -> T:
+        if self._size == 0:
+            raise Exception('No values in stack to remove')
+        # use second stack to extract item of interest
+        data: T = self._extract_first_in('pop')
+        self._size -= 1
+        return data
+
+    def peek(self) -> T:
+        if self._size == 0:
+            raise Exception('No values in stack to remove')
+        return self._extract_first_in('peek')
+
+    def is_empty(self) -> bool:
+        return self._size == 0
+    
+    def _extract_first_in(self, operation: str) -> T:
```

Comment:
in cases like this, it's preferable to use [`enum.Enum`](https://docs.python.org/3/library/enum.html#creating-an-enum) to make the options type-safe.

it may be even more preferable to split this in two functions (one to pop and another one to peek) that call other helper functions to do 1 and 3 (and in fact can be done with a single helper function where the stacks are passed as parameters). that way the enum is not even needed.

Reply:
Split into two functions here:  4e38eabef4b023618802ec82779d57318bf16a47